<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A Swift, Versatile Phylogenomic and High-Throughput Sequencing Simulator • jackalope</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A Swift, Versatile Phylogenomic and High-Throughput Sequencing Simulator">
<meta property="og:description" content="Simply and efficiently
    simulates (i) variants from reference genomes and (ii) reads from both Illumina 
    &lt;https://www.illumina.com/&gt;
    and Pacific Biosciences (PacBio) &lt;https://www.pacb.com/&gt; platforms. 
    It can either read reference genomes from FASTA files or simulate new ones.
    Genomic variants can be simulated using summary statistics, phylogenies, 
    Variant Call Format (VCF) files, and coalescent simulations—the latter of which
    can include selection, recombination, and demographic fluctuations.
    'jackalope' can simulate single, paired-end, or mate-pair Illumina reads, 
    as well as PacBio reads.
    These simulations include sequencing errors, mapping qualities, multiplexing,
    and optical/polymerase chain reaction (PCR) duplicates.
    Simulating Illumina sequencing is based on ART
    by Huang et al. (2012) &lt;doi:10.1093/bioinformatics/btr708&gt;.
    PacBio sequencing simulation is based on 
    SimLoRD  by Stöcker et al. (2016) &lt;doi:10.1093/bioinformatics/btw286&gt;.
    All outputs can be written to standard file formats.">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">jackalope</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/jackalope-intro.html">Intro to jackalope</a>
    </li>
    <li>
      <a href="articles/sub-models.html">Models of nucleotide substitution</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lucasnell/jackalope">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<!-- README.md is generated from README.Rmd. Please edit that file -->

<div id="jackalope" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#jackalope" class="anchor"></a>jackalope <img src="reference/figures/logo.png" align="right" alt="" width="120">
</h1></div>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>For studies using high-throughput sequencing (HTS) data, simulations can be vital for planning sampling design and testing bioinformatic tools. However, most HTS sequencing tools provide only very simple ways of adding deviations from a reference genome. For HTS studies that focus on patterns of genomic variation among individuals, populations, or species, having a tool that can simulate realistic patterns of molecular evolution and generate HTS data from those simulations would be quite useful.</p>
<p><code>jackalope</code> simply and efficiently simulates (i) variants from reference genomes and (ii) reads from both Illumina and Pacific Biosciences (PacBio) platforms. It can either read reference genomes from FASTA files or simulate new ones. Genomic variants can be simulated using summary statistics, phylogenies, Variant Call Format (VCF) files, and coalescent simulations—the latter of which can include selection, recombination, and demographic fluctuations. <code>jackalope</code> can simulate single, paired-end, or mate-pair Illumina reads, as well as reads from Pacific Biosciences These simulations include sequencing errors, mapping qualities, multiplexing, and optical/PCR duplicates. All outputs can be written to standard file formats.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># To install the latest stable version from CRAN:</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"jackalope"</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co"># Or the the development version from GitHub:</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co"># install.packages("devtools")</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"lucasnell/jackalope"</span>)</a></code></pre></div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>Below shows how to simulate a 10kb genome, then create variants from that genome using a phylogenetic tree:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(jackalope)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">reference &lt;-<span class="st"> </span><span class="kw"><a href="reference/create_genome.html">create_genome</a></span>(<span class="dt">n_seqs =</span> <span class="dv">10</span>, <span class="dt">len_mean =</span> <span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">tr &lt;-<span class="st"> </span>ape<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ape/topics/rtree">rcoal</a></span>(<span class="dv">5</span>)</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">ref_variants &lt;-<span class="st"> </span><span class="kw"><a href="reference/create_variants.html">create_variants</a></span>(reference, <span class="kw"><a href="reference/vars_phylo.html">vars_phylo</a></span>(tr), <span class="kw"><a href="reference/sub_models.html">sub_JC69</a></span>(<span class="fl">0.1</span>))</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">ref_variants</a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co">#&gt;                            &lt;&lt; Variants object &gt;&gt;</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co">#&gt; # Variants: 5</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="co">#&gt; # Mutations: 17,679</span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="co">#&gt;                         &lt;&lt; Reference genome info: &gt;&gt;</span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="co">#&gt; &lt; Set of 10 sequences &gt;</span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="co">#&gt; # Total size: 10,000 bp</span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="co">#&gt;   name                          sequence                             length</span></a>
<a class="sourceLine" id="cb2-14" data-line-number="14"><span class="co">#&gt; seq0       CTGGCATTGAATCATATGAGGTGGC...GTTGCACGATTGATTAAATTCCTGAA      1000</span></a>
<a class="sourceLine" id="cb2-15" data-line-number="15"><span class="co">#&gt; seq1       CACTCCGTCGCACACTAGGTTTCGA...GAGCTCGCGTACATGGAGCATTCTGT      1000</span></a>
<a class="sourceLine" id="cb2-16" data-line-number="16"><span class="co">#&gt; seq2       CTTAGCCGGAGCGACTCGGAGCAAC...GCGTAATATGCCAGGTCCCGCGTGGC      1000</span></a>
<a class="sourceLine" id="cb2-17" data-line-number="17"><span class="co">#&gt; seq3       CGCCTTCCATTTAGGACTTGTATTG...TAAACTCCATGTGACTGTAATGTCAG      1000</span></a>
<a class="sourceLine" id="cb2-18" data-line-number="18"><span class="co">#&gt; seq4       GGGTGATATGGTGTGCATGCTGAAT...AGTCTAGAGTCTCTGGGAGGTCAGGT      1000</span></a>
<a class="sourceLine" id="cb2-19" data-line-number="19"><span class="co">#&gt; seq5       TTCGTTGGTGGGTGTCCTATGCTAC...CCCGCCGGTTTGACTTACTCGATTGG      1000</span></a>
<a class="sourceLine" id="cb2-20" data-line-number="20"><span class="co">#&gt; seq6       GCATGGACAGATGTGATCTGAGTAT...GACCCCATAAGGCCTGGGACACTGTG      1000</span></a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="co">#&gt; seq7       TCGTTTCAACGTCCTTAAGTGTAGT...CTCGTTAGCTCTCCGAGGAGACGAGG      1000</span></a>
<a class="sourceLine" id="cb2-22" data-line-number="22"><span class="co">#&gt; seq8       CAGGTAAGTTATCAAAGAACCTTCC...GCATCACCTCGCAAGGAGACTCGTTA      1000</span></a>
<a class="sourceLine" id="cb2-23" data-line-number="23"><span class="co">#&gt; seq9       GGTAGTAATTAGGCTTAAAATAGCA...AACAAATGTTCGGCATACGATCTACG      1000</span></a></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/lucasnell/jackalope">https://​github.com/​lucasnell/​jackalope</a>
</li>
<li>Report a bug at <br><a href="https://github.com/lucasnell/jackalope/issues">https://​github.com/​lucasnell/​jackalope/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Lucas A. Nell <br><small class="roles"> Copyright holder, author, maintainer </small> <a href="https://orcid.org/0000-0003-3209-0517" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
</ul>
</div>

      <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.repostatus.org/#active"><img src="https://www.repostatus.org/badges/latest/active.svg" alt="Project Status: Active – The project has reached a stable, usable state and is being actively developed."></a></li>
<li><a href="https://travis-ci.com/lucasnell/jackalope"><img src="https://travis-ci.com/lucasnell/jackalope.svg?branch=master" alt="Build Status"></a></li>
<li><a href="https://codecov.io/gh/lucasnell/jackalope"><img src="https://codecov.io/gh/lucasnell/jackalope/branch/master/graph/badge.svg" alt="codecov"></a></li>
<li><a href="https://ci.appveyor.com/project/lucasnell/jackalope"><img src="https://ci.appveyor.com/api/projects/status/github/lucasnell/jackalope?branch=master&amp;svg=true" alt="AppVeyor build status"></a></li>
</ul>
</div>
</div>

</div>


      <footer><div class="copyright">
  <p>Developed by Lucas A. Nell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
